sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala

```diff
@@ -572,6 +572,29 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
     assert(df.schema.map(_.name) === Seq("value"))
   }
 
+  test("SPARK-28189 drop column using drop with column reference with case-insensitive names") {
+    // With SQL config caseSensitive OFF, case insensitive column name should work
+    withSQLConf(SQLConf.CASE_SENSITIVE.key -> "false") {
+      val col1 = testData("KEY")
+      val df1 = testData.drop(col1)
+      checkAnswer(df1, testData.selectExpr("value"))
+      assert(df1.schema.map(_.name) === Seq("value"))
+
+      val col2 = testData("Key")
+      val df2 = testData.drop(col2)
+      checkAnswer(df2, testData.selectExpr("value"))
+      assert(df2.schema.map(_.name) === Seq("value"))
+    }
+
+    // With SQL config caseSensitive ON, AnalysisException should be thrown
+    withSQLConf(SQLConf.CASE_SENSITIVE.key -> "true") {
+      val e = intercept[AnalysisException] {
+        testData("KEY")
```

Comment:
wait, it seems this test is not related to this pr?
you just wanted to do like this?
```
  test("SPARK-28189 drop column using drop with column reference with case-insensitive names") {
    var caseInsensitiveCol: Column = null
    withSQLConf(SQLConf.CASE_SENSITIVE.key -> "false") {
      caseInsensitiveCol = testData("KEY")
    }

    Seq("true", "false").foreach { isCaseSensitive =>
      withSQLConf(SQLConf.CASE_SENSITIVE.key -> isCaseSensitive) {
        caseInsensitiveCol = testData("KEY")
        val df = testData.drop(caseInsensitiveCol)
        checkAnswer(df, testData.selectExpr("value"))
        assert(df.schema.map(_.name) === Seq("value"))
      }
    }
  }
```

Reply:
? @maropu . In your example, the following will fail in the same way because `testData` only has `key` and `value`.
```
      withSQLConf(SQLConf.CASE_SENSITIVE.key -> isCaseSensitive) {
        caseInsensitiveCol = testData("KEY")
```
For me, the test case looked okay because `caseSensitive` mode doesn't allow that kind of unmatched column from the beginning.

Reply:
oh, sorry and my bad. That line wasn't needed and I updated the example above.
Anyway, I just a bit worry about the behaivour change in the code flow below;
```
scala> val testData = Seq(("a", 1)).toDF("key", "value")

scala> sql("SET spark.sql.caseSensitive=false")
scala> val caseInsensitiveCol = testData("KEY")

scala> sql("SET spark.sql.caseSensitive=true")
scala> testData.drop(caseInsensitiveCol).show()
// v2.4.3
+---+-----+
|key|value|
+---+-----+
|  a|    1|
+---+-----+

// master w/this pr
+-----+
|value|
+-----+
|    1|
+-----+
```

Reply:
First, it's not a correct(?) use-case to switch the case sensitivity between declaring the variable and using the variable. 
Second, I merged this to `master` only with the similar concern. :)
Since this is a bug fix, I think we don't need a documentation about this. And, 3.0.0 is a good place to change those behavior change bug fix.

Reply:
ok, thanks for the check!

Reply:
Do you have any suggestion for your concern? Then, please share it with us~ You're welcome.

Reply:
NVM. I just a bit worry about not use cases but the test coverage.

Reply:
Yes. This test case of the second part (SQLConf.CASE_SENSITIVE.key -> "true") is weird to me.

Reply:
Anyway, we can keep it unchanged. Ideally, we do not need the second part.

Reply:
ok, I made a pr to drop that part: https://github.com/apache/spark/pull/25216
anyway, thanks for the check!